File: core/src/vct/col/ast/TypeVariable.java
package vct.col.ast;

import java.util.Arrays;

public class TypeVariable extends Type {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  public final String name;
  
  public TypeVariable(String name) {
    this.name=name;
  }
  
  public boolean isNumeric() {
    return false;
  }

  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
 
  @Override
  public <T> T accept_simple(TypeMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }


  @Override
  public boolean supertypeof(ProgramUnit context, Type t) {
    return false;
  }
  
  public int hashCode(){
    return name.hashCode();
  }
  
  public boolean equals(Object o){
    if (o instanceof TypeVariable){
      return ((TypeVariable)o).name.equals(name);
    }
    return false;
  }

}

File: vercors/src/main/java/vct/col/ast/type/TypeVariable.scala
package vct.col.ast.`type`

import vct.col.ast.stmt.decl.ProgramUnit
import vct.col.ast.util.{ASTVisitor, TypeMapping}
import vct.col.util.{ASTMapping, ASTMapping1, VisitorHelper}

case class TypeVariable(val name:String) extends Type with VisitorHelper {
  override def isNumeric() = false
  override def hashCode() = name.hashCode();
  override def supertypeof(context:ProgramUnit, t:Type) = false

  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
  override def accept_simple[T](m:TypeMapping[T]) = handle_standard(() => m.map(this))

  override def equals(o:Any) = o match {
    case tv:TypeVariable => tv.name.equals(this.name)
    case _ => false
  }

  override def debugTreeChildrenFields: Iterable[String] = Seq("args")
  override def debugTreePropertyFields: Iterable[String] = Seq("name")
}
